### ISSUE TITLE
KeyError `id` when creating an Account

### ISSUE THREAD

**Issue by rocketrip-al** [contextio/Python-ContextIO#35]
I'm running on Python 3.5 and am running into issues creating an `Account` by `id`.

```python
cio = contextio.ContextIO(
    consumer_key=KEY,
    consumer_secret=SECRET
)
contextio.Account(cio, {'id': ACCOUNT_ID})
```

This produces this error:
```
  File ".../lib/python3.5/site-packages/contextio/__init__.py", line 702, in __init__
    super(Account, self).__init__(parent, 'accounts/{id}', defn)
  File ".../lib/python3.5/site-packages/contextio/__init__.py", line 598, in __init__
    self.base_uri = quote(base_uri.format(**unidict))
KeyError: 'id'
```

So the question is why do we convert to bytes if we have a string here?  Is this necessary for earlier versions of python?  It works if I don't decode to bytes.
https://github.com/contextio/Python-ContextIO/blob/master/contextio/__init__.py#L597

Or is there something else I should be doing?

Thanks for the help!

**Comment by Daenyth** [comment]
I'm seeing a similar thing with Message.

```python
contextio.Message(None, {'message_id':'mid'})
# KeyError: 'message_id'
```

I can pass bytes for now;
```
In [4]: contextio.Message(None, {b'message_id':'mid'})
Out[4]: <contextio.Message at 0x7fb2e0737080>
```

---

In python2 this probably didn't matter, because a map key of either type can look up the other as long as the unicode can encode to ascii. It matters quite a bit now.

**Comment by Daenyth** [comment]
I believe that changing this line

```python
unidict = {k.encode('utf8') if isinstance(k, six.string_types) else k: v.encode('utf8') if isinstance(v, six.string_types) else v for k, v in list(defn.items())}
```

To this:

```python
# unidict is a map from unicode keys to unicode values
unidict = {six.text_type(k): six.text_type(v) for k, v in defn.items()}
```

Should work for instantiation.

Handling the instantiation like that might require that other parts of the module have to `encode` from unicode to bytes.

I can put in a PR but because this project is not tested I have no idea if I might break something.

**Comment by Daenyth** [comment]
I have an idea for another approach, I'll try writing a Mapping class that provides the same api as dict but permits you to look up keys with weak typing

**Comment by Daenyth** [comment]
Wrote it, I'll be testing it with contextio shortly. https://gist.github.com/Daenyth/4bcdd114861a19f14f2c

**Comment by Daenyth** [comment]
Using this seems to work so far, along with passing in the naive dict. I'll report back again when I have some field testing 

```
# We monkeypatch Resource.__init__ because it is incorrect under python3.
# See https://github.com/contextio/Python-ContextIO/issues/35
def resource_init(self, parent, base_uri, defn):
    """Constructor."""

    if defn == "":
        logging.error('Empty response received for ' + base_uri + "")
        return

    try:
        defn = contextio.uncamelize(defn)
    except:
        logging.error('Invalid response received for ' + base_uri + "")
        return

    for k in self.__class__.keys:
        if k in defn:
            setattr(self, k, defn[k])
        else:
            setattr(self, k, None)

    self.parent = parent
    # This is the line causing most of the bug --Gavin
    #unidict = {k.encode('utf8') if isinstance(k, six.string_types) else k: v.encode('utf8') if isinstance(v, six.string_types) else v for k, v in list(defn.items())}
    unidict = defn
    self.base_uri = contextio.quote(base_uri.format(**unidict))

contextio.Resource.__init__ = resource_init
```

**Comment by rocketrip-al** [comment]
Thanks for the reply @Daenyth.  I think I understand the issue a little bit better now.  Coincidentally, I applied a local patch that does what you did in the previous comment (using the unaltered `defn`).

Since the two cases are fundamentally different, what do you think about having separate code for each python version?  It would be more explicit, albeit a little uglier.

```python
if six.PY2:
    unidict = {k.encode('utf8') if isinstance(k, six.string_types) else k: v.encode('utf8') if isinstance(v, six.string_types) else v for k, v in list(defn.items())}
else:
    unidict = defn
```

**Comment by Daenyth** [comment]
I don't think it's necessary to have two cases; you can express the same semantics in either version. In fact I think the best approach is to coerce all keys to be unicode instead of bytes, and internally use unicode to do lookups. This would work fine on both versions

**Comment by rocketrip-al** [comment]
Yeah I agree.  I was just wary of why we were encoding to bytes here in the first place.  So you're saying we should just do this:

```
unidict = {k.decode('utf8') if isinstance(k, bytes) else k: v.decode('utf8') if isinstance(v, bytes) else v for k, v in list(defn.items())}
```

**Comment by Daenyth** [comment]
That's what I'd recommend.

**Comment by sushithegreat** [comment]
Should be fixed in 1.9
